test(native): Add `AbstractTestQueries` to native tests

pramodsatya opened the pull request:
## Description
Adds support to run testcases from `AbstractTestQueries` with the C++ worker in `presto-native-tests`. These testcases are run with Presto's iterative optimizer disabled in `TestNonIterativeDistributedQueries` and with the default native hive query runner in `TestTpchDistributedQueries`. 

## Motivation and Context
Improves Presto C++ e2e test coverage with sidecar by extending production tests from `presto-tests` module to use the native worker.

## Release Notes

```
== RELEASE NOTES ==

General Changes
* Add testcases from `AbstractTestQueries` to `presto-native-tests`.
```

steveburnett reviewed (dismissed):
LGTM! (docs)

Reviewed README.md, no concerns. Thanks for the doc!

tdcmeehan commented:
Are these tests being run as part of the CI?

pramodsatya commented:
They are not being run as part of CI currently, but we plan to setup a pipeline to run these tests on a regular cadence subsequently. Will also add a CI job to run these tests in a follow-up PR.

jaystarshot commented on presto-native-tests/src/test/java/com.facebook.presto.nativetests/TestAggregations.java:
```diff
@@ -0,0 +1,43 @@
+public class TestAggregations
+        extends AbstractTestAggregationsNative
+{
+    private static final String storageFormat = "PARQUET";
```
This is just testing parquet

pramodsatya replied:
Yes the intention was to primarily test with parquet files. Most of the current e2e tests in `presto-native-execution` test with DWRF format. We can extend this to test with other file formats going forward.

pramodsatya replied:
> This is just testing parquet

Tests with `DWRF` are added here: https://github.com/prestodb/presto/pull/24809.

pramodsatya commented:
> Just a suggestions for the release note: Consider using the description of the PR in the release note entry to help a release note reader understand what this module does.
> 
> ```
> == RELEASE NOTES ==
> General Changes
> * Add presto-native-tests module to run end-to-end tests with Presto native workers. :pr:`23671`
> ```

Thanks for the suggestion @steveburnett, updated accordingly.

zuyu commented:
Fyi, build error on your branch, after resolving conflicts w/ the latest `master` branch.

```
...
[INFO] presto-native-tests 0.290-SNAPSHOT ................. FAILURE [  7.832 s]
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  19.526 s (Wall Clock)
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (checkstyle) on project presto-native-tests: Failed during checkstyle execution: Unable to find configuration file at location: src/checkstyle/presto-checks.xml: Could not find resource 'src/checkstyle/presto-checks.xml'. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :presto-native-tests
```

aditi-pandit reviewed (commented):
@pramodsatya @soumiiow : Can we move the custom functions test to a separate PR ? Don't think it is depending on any other changes in this PR is it ?

pramodsatya commented:
> @pramodsatya @soumiiow : Can we move the custom functions test to a separate PR ? Don't think it is depending on any other changes in this PR is it ?

@aditi-pandit, the concern in moving this to a separate PR (and test class) is that there are 41 other test classes in different modules that run the testcases from `AbstractTestQueries`, which include the e2e tests for Presto custom functions. So we did not want to break down `AbstractTestQueries`. Would it be fine to keep it in this PR?

pramodsatya commented:
@aditi-pandit, I missed that `AbstractTestCustomFunctions` was a part of the custom function tests added here and this wasn't the intention. I have reverted this now and moved custom function tests back to `AbstractTestQueries`. Could you please take another look at the custom function tests commit?

pramodsatya commented:
> @pramodsatya : I would be fine to create a separate test class 'AbstractTestCustomFunctions' with repetition for presto-native-tests only.
> 
> > Thanks for the suggestion @aditi-pandit, we went with this approach initially by moving the custom udf testcases to a separate test class, `AbstractTestCustomFunctions` (prototyped [here](https://github.com/pramodsatya/presto/commits/dyfunc_native_tests/)). However, we were concerned moving these testcases to `AbstractTestCustomFunctions` would result in them not being run by the other test suites that derive from `AbstractTestQueries`. We could duplicate the custom udf tests and copy them from `AbstractTestQueries` to `AbstractTestCustomFunctions` instead of moving them, not sure if that would be preferable as it would result in code duplication. Could you please share how to proceed?

Thanks @aditi-pandit, opened a separate PR for custom function native tests: https://github.com/prestodb/presto/pull/25480. Could you please take a look?

pramodsatya commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
Thanks for the suggestion @aditi-pandit, moved these changes to a separate PR: https://github.com/prestodb/presto/pull/25486. Could you please help review this change?

aditi-pandit commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/AbstractTestQueriesNative.java:
```diff
@@ -0,0 +1,1174 @@
+    /// https://github.com/prestodb/presto/issues/24402.
+    @Override
+    @Test(enabled = false)
+    public void testArrayCumSumDecimals() {}
```
I have seen a PR from Joe for this. It will be good to enable this test and assert the error messages so that he will have to explicitly make changes when his PR makes it to Velox.

aditi-pandit commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/AbstractTestQueriesNative.java:
```diff
@@ -0,0 +1,1174 @@
+    /// Planner error: type of variable 'expr' is expected to be varchar(5), but the actual type is varchar.
+    @Override
+    @Test(enabled = false)
+    public void testCaseToMapOptimization() {}
```
Check error messages in test instead.

aditi-pandit commented on presto-docs/src/main/sphinx/presto_cpp/limitations.rst:
You could move the changes in limitations.rst and TestNonIterativeDistributedQueries and TestTpchDistributedQueries into a separate PR as that will be much easier to review and submit than this PR.

aditi-pandit commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/NativeQueryRunnerUtils.java:
```diff
@@ -282,7 +282,7 @@ public static void createCustomer(Session session, QueryRunner queryRunner)
         if (!queryRunner.tableExists(session, "customer")) {
             queryRunner.execute(session, "CREATE TABLE customer AS " +
-                    "SELECT custkey, name, address, nationkey, phone, acctbal, comment, mktsegment " +
+                    "SELECT custkey, name, address, nationkey, phone, acctbal, mktsegment, comment " +
```
I don't follow why this fixed something. Please can you elaborate.

aditi-pandit commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/NativeTestsUtils.java:
```diff
@@ -47,10 +47,10 @@ public static void createTables(String storageFormat)
                     .build();
             if (storageFormat.equals("DWRF")) {
-                NativeQueryRunnerUtils.createAllTables(javaQueryRunner, true);
+                NativeQueryRunnerUtils.createAllTablesStandard(javaQueryRunner, true);
```
Don't understand why you changed this to standard.. Seems non-intuitive as the calling method is called createTables.

aditi-pandit commented on presto-main-base/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java:
```diff
@@ -480,7 +482,16 @@ public static boolean isSupportedArrayContainsFilter(FunctionResolution function
                 RowExpression array = callExpression.getArguments().get(0);
                 RowExpression element = callExpression.getArguments().get(1);
-                checkState(array.getType() instanceof ArrayType && ((ArrayType) array.getType()).getElementType().equals(element.getType()));
+                // varchar(n) is not supported in Presto C++, compare base types of array and element for varchar to
```
Can you move this commit outside of this PR and write a separate individual test for it.  You can start with the original test you fixed, and then we might want to add more variations to it.

aditi-pandit commented on presto-main-base/src/main/java/com/facebook/presto/metadata/FunctionSignatureMatcher.java:
```diff
@@ -71,7 +71,7 @@ public Optional<Signature> match(Collection<? extends SqlFunction> candidates, L
                 .collect(Collectors.toList());
 
-        match = matchFunctionExact(genericCandidates, parameterTypes);
+        match = matchFunctionGeneric(genericCandidates, parameterTypes);
```
Separate this into a PR also for just this condition and write an individual test for it.

aditi-pandit reviewed (commented):
@pramodsatya : Thanks for the iterations of this PR. Have the next round of comments.

aditi-pandit commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/AbstractTestQueriesNative.java:
```diff
@@ -0,0 +1,1174 @@
+    /// Actual   :[5, 5, F]
+    @Override
+    @Test(enabled = false)
+    public void testTopNUnpartitionedWindow() {}
```
Create an issue for this.

steveburnett commented:
Suggested changes for the release note. 

```
== RELEASE NOTES ==

Prestissimo (Native Execution) Changes
* Add testcases from ``AbstractTestQueries`` to ``presto-native-tests``.
```

pramodsatya committed:
refactor(testing): Refactor testcases for native-tests

pramodsatya reviewed (commented):
Thanks for the feedback @aditi.pandit, addressed the comments. Could you PTAL?

pramodsatya commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/TestTpchDistributedQueries.java:
```diff
@@ -0,0 +1,118 @@
+    @Test
+    public void testAnalyzePropertiesSystemTable()
+    {
+        assertQuery("SELECT COUNT(*) FROM system.metadata.analyze_properties WHERE catalog_name = 'tpch'", "SELECT 0");
```
@aditi-pandit, this does not fail on native, it is added here to maintain consistency with the `TestTpchDistributedQueries` from `presto-tests`:
https://github.com/prestodb/presto/blob/9ebd1f2a97256b9f0e5e77c5722bd39045034778/presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedQueries.java#L43

Ideally the tests here should be moved to a separate abstract class, I can take this up in a subsequent PR?

aditi-pandit replied:
Yeah... you can move the tests to a separate abstract class. Lets avoid rewriting the tests.

aditi-pandit commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeIcebergTpchQueries.java:
```diff
@@ -32,7 +32,7 @@ public abstract class AbstractTestNativeIcebergTpchQueries
     {
         QueryRunner queryRunner = (QueryRunner) getExpectedQueryRunner();
-        createLineitemStandard(queryRunner);
+        createLineitemStandard(queryRunner, true);
```
The casting should depend on storageFormat as some types like DWRF don't support DATE, but Parquet does. Can you check ?

aditi-pandit reviewed (commented):
@pramodsatya : Another high level comment: Whether we cast Date depends on the storage format as DWRF doesn't support DATE but Parquet does. Ideally we should use that to guide the casting and avoid the extra parameter as much as possible. Lets cleanup guided by that.

aditi-pandit commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/AbstractTestQueriesNative.java:
```diff
@@ -0,0 +1,1158 @@
+        assertQueryFails("SELECT greatest(rgb(255, 0, 0))", errorMessage);
+    }
+
+    /// The output JSON formatted string is different in Presto C++.
```
@amitkdutta : Can you review these tests for json_format usage ?

pramodsatya commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeHiveExternalTableTpchQueries.java:
```diff
@@ -109,7 +109,7 @@ protected void createTables()
         Session session = Session.builder(super.getSession()).setCatalog(HIVE).setSchema(TPCH_SCHEMA).build();
         createOrders(session, javaQueryRunner, true);
-        createLineitemStandard(session, javaQueryRunner);
+        createLineitemStandard(session, javaQueryRunner, true);
```
Reverted, thanks.

pramodsatya committed:
test(native): Add AbstractTestQueries to native-tests

pramodsatya reviewed (commented):
Thanks @aditi-pandit, I have reverted the `castDateToVarchar` parameter to use `storageFormat` instead, and separated this change to https://github.com/prestodb/presto/pull/26955.

Could you please take another look?

aditi-pandit commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/AbstractTestQueriesNative.java:
```diff
@@ -0,0 +1,1172 @@
+
+    @Override
+    @Test
+    public void testMinMaxByToWindowFunction()
```
Why are these rewrites needed ?

aditi-pandit commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/AbstractTestQueriesNative.java:
```diff
@@ -0,0 +1,1172 @@
+    /// The error message for the case where subquery returns multiple rows differs in Presto and Velox.
+    @Override
+    @Test
+    public void testScalarSubquery()
```
Is there a way to avoid this test repetition ? It might be okay to refactor testScalarSubquery with a common function that takes the error strings as a parameter.

aditi-pandit commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/AbstractTestQueriesNative.java:
```diff
@@ -0,0 +1,1172 @@
+    }
+
+    /// TODO: Check why Presto C++ doesn't throw an error for certain queries.
+    @Override
```
Lets get into the details of this. Please create issues if needed.

aditi-pandit commented on presto-native-tests/src/test/java/com/facebook/presto/nativetests/AbstractTestQueriesNative.java:
```diff
@@ -0,0 +1,1172 @@
+        }
+    }
+
+    /// TODO: The last query in this testcase is flaky in Presto C++.
```
Can we investigate the flakiness ? Please create an issue for this.

aditi-pandit reviewed (commented):
Thanks @pramodsatya. Have bunch of comments.

pramodsatya committed:
Address review comments
